### ISSUE TITLE
discussion: porting Halide to nim

### ISSUE THREAD

**Issue by timotheecour** [mratsim/Arraymancer#347]
opening this issue to track ideas for how to port Halide to nim.
This could be either using halide in Nim via:
* A1: a wrapper (eg using nimterop) or
* A2: re-implementing some ideas in nim directly, which although a lot of work would result in a much simpler solution given nim's better DSL abilities

## proposal: approach 1
use a plugin at CT approach as done in nimterop (https://github.com/genotrance/nimterop/pull/66) to transparently compile (using halide) a c++/Halide DSL, save to a shared library and cache it; eg:

```nim
# this is in Nim

import halide_funs

cPluginHalide: """ // this is compiled on the fly and cached, using a hash of the string, eg saved to: /tmp/nimhalide/libplugin_12abf23.dylib
// this is C++ halide code
Func blur_3x3(Func input) { // example from https://en.wikipedia.org/wiki/Halide_(programming_language)
  Func blur_x, blur_y;
  Var x, y, xi, yi;

  // The algorithm - no storage or order
  blur_x(x, y) = (input(x-1, y) + input(x, y) + input(x+1, y))/3;
  blur_y(x, y) = (blur_x(x, y-1) + blur_x(x, y) + blur_x(x, y+1))/3;

  // The schedule - defines order, locality; implies storage
  blur_y.tile(x, y, xi, yi, 256, 32)
        .vectorize(xi, 8).parallel(y);
  blur_x.compute_at(blur_y, x).vectorize(x, 8);

  return blur_y;
}"""

proc main() =
  let image = loadImage("img.jpg")
  blur_3x3(image)
```

## [EDIT] proposal: approach 2
using https://github.com/fragcolor-xyz/nimline could be another possiblity, but would probalby need adaptation as halide compiler needs to be run on C++/halide program 

## related
* halide: https://en.wikipedia.org/wiki/Halide_(programming_language)
* some discussion on gitter https://gitter.im/Arraymancer/Lobby?at=5c4f1b7d13a2814df6ded8f5 + a few hours before that

* https://github.com/mratsim/Arraymancer/issues/203 briefly mentions halide but wasn't the focus of that issue

* other language integrations:
  * python: https://stackoverflow.com/questions/34439657/halide-with-c-layout-numpy-arrays
  * rust: [Rust Image Processing DSL – dpzmick.com](http://dpzmick.com/2016/08/11/rust-jit-image-processing/)

* tutorial http://halide-lang.org/tutorials/tutorial_lesson_16_rgb_generate.html

* nlvm: https://github.com/arnetheduck/nlvm

**Comment by mratsim** [comment]
As discussed, this sounds like a very interesting idea and from a learning perspective I'd rather implement it from scratch, just like Arraymancer started as a way to learn NN from scratch instead of reusing Tensorflow or Torch.

## Overview of Halide

Halide is a C++ domain specific language for computational image processing. It started with the idea of separating the algorithm (the what) from its efficient hardware implementation (the how). This is similar to a (functional programming) source file vs machine code when passed through an optimizing compiler.

The "how" can be hinted to assist the compiler in its threading and vectorization by specifying cache blocking parameters, unrolling factor and threading hints.

Furthermore Halide supports multiple backends (x86, ARM, Vulkan, DX12, OpenGL, OpenCL, Cuda ...) including a JIT backend that will dispatch with the latest instruction set supported (for example AVX512).

The pipeline is: C++ DSL --> Halide Intermediate Representation (Computational Graph) --> LLVM (AOT or JIT) --> machine code.

## Advantages

### In general

- Short code, much faster to develop new efficient algorithms.
- No need to fiddle with platform specific optimisations like SIMD.
- Portability
- Computation can be fused across functions and more parallelism can be exposed.
- Much easier to test different parallel schedule

### For ML and DL

The main advantages are the portability with SIMD optimisations opportunities and computation fusion to prevent intermediate allocation and results.

Parallelism in ML/DL is either straightforward (it's a single loop) or very complex and would need manual involvement:

- For example Halide is memory-bound on matrix multiplication (0.5~1 TFLOP/s at most while I can reach [1.8TFLOP/s with OpenBLAS or MKLDNN](https://github.com/numforge/laser/blob/56df4643530ada0a01d29116feb626d93ac11379/benchmarks/gemm/gemm_bench_float32.nim#L379-L465))
- Halide is unsuited for trees ensembles like random forests

Some DL layers that would benefits from Halide-like functionality:
  - Convolution (apparently Halide has a schedule that reaches MKL-DNN performance)
  - Locally-Connected Layer/Unshared Convolution (#255). Though this might be faster with summed-area table/integral image and then a simple for loop.
  - Bilinear
  - Pooling layers (Max, Average, Region of Interest Pooling)

### Disadvantages

1. Halide is pretty much a static graph compiler like Tensorflow XLA which is not that great for research and debugging. It's a pain to use Tensorflow for RNNs and GANs for example.

    The main issue is the lack of control flow like for loop and if/else branches.

2.  Another disadvantage is the LLVM dependency at runtime, but that only concerns JIT. LLVM is a huge library.

### Beyond Halide

Halide has been expanded with autodiff capabilities in [gradient-halide](https://people.csail.mit.edu/tzumao/gradient_halide/) which may greatly enhance autograd which is a recurrent source of slowness in Arraymancer due to the need of a DAG which doesn't play well with reference counting. However while gradient-halide results are promising, TVM (a fork of Halide for DL) [reported perf issues](https://discuss.tvm.ai/t/automatic-differentiation-on-the-level-of-tvm-ir/781)

### Related projects
(To be edited)

- Tensorflow XLA: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/g3doc/overview.md
- PyTorch Glow: https://github.com/pytorch/glow
- Gradient Halide: https://people.csail.mit.edu/tzumao/gradient_halide/
- OpenAI Dali: http://learningsys.org/nips18/assets/papers/100CameraReadySubmissionDali__Lazy_Compilation_of_Dynamic_Computation_Graphs.pdf
- DLMC TVM: https://github.com/dmlc/tvm
- Intel PlaidML: https://github.com/plaidml/plaidml
- The Tensor Algebra Compiler: https://github.com/tensor-compiler/taco

TODO - approaches to porting.

**Comment by mratsim** [comment]
I find the need to go through a text file quite clunky and slow in the long term.

Since LLVM has a [bit code representation ](https://llvm.org/docs/BitCodeFormat.html) of its IR I was thinking of targeting that directly using bindings akin to [LLVM-D](https://github.com/MoritzMaxeiner/llvm-d).

This is also [what Halide is doing.](https://github.com/halide/Halide/blob/6169020460efed0cff6973d4c29e81c1ceedccdc/src/CodeGen_Internal.h#L94-L98).

From an interface point of view, I probably can re-use the JIT Assembler for x86-64 [I created here](https://github.com/numforge/laser/tree/27e4a77c7808ac5b5dcee08ddf3baa320aed47c8/laser/photon_jit/x86_64) except that instead of assembling x86 I'm assembling LLVM IR.

Alternatively, it seems like we may be able [to directly JIT an AST representation](https://llvm.org/docs/tutorial/BuildingAJIT4.html) without having to go through LLVM IR phase.

**Comment by mratsim** [comment]
Some advances in my research (will find a place to store that in the future, probably in [Laser](https://github.com/numforge/laser))

# Code generation

- Kai Nacke, maintainer of LDC (LLVM D Compiler), did a very nice talk about what he learned by implementing his own compiler from scratch at FOSDEM 2019. [Slides](https://fosdem.org/2019/schedule/event/llvm_irgen/attachments/paper/3170/export/events/attachments/llvm_irgen/paper/3170/FOSDEM_2019_slides.pdf).

# Alternative HPC compiler

- http://okmij.org/ftp/meta-programming/HPC.html and the Shonan challenge, several HPC problems with a known tuned manual solution that would be nice to automatically optimize: 
    - Paper: http://okmij.org/ftp/meta-programming/Shonan-challenge.pdf
    - Report: http://okmij.org/ftp/meta-programming/shonan2-report.pdf
    - 1D Conv optimisation in OCaml: http://okmij.org/ftp/meta-programming/lift/convolution.ml

# Semantic pass and the extension problem

Halide and PyTorch Glow are using the visitor pattern which looks to me like a workaround for a not expressive enough language.

We need to be able to extend functions without touching the core ones and if possible data types as well.

For example we could have the functions `Add` and `Mul` defined on `Tensor` and `CudaTensor` today and a user library implement `Div` and also extend `Add` and `Mul` on `VulkanTensor`.

Data types are probably not that critical, i.e. we could use ADTs + Generics for Tensor[float32], VulkanTensor[float16], ...

In research, there are several approaches to handle that:

  - Tagless Final, embedding DSL into a typed language: http://okmij.org/ftp/tagless-final/
  - Object Algebra: http://ponies.io/posts/2015-07-15-solving-the-expression-problem-in-python-object-algebras-and-mypy-static-types.html
  - Attribute Grammar (needs lazy eval): 
        - https://www.reddit.com/r/haskell/comments/6ugzhe/walking_the_ast_without_the_visitor_pattern/
        - https://play.rust-lang.org/?gist=1f7e5a01ec3d0010c50ecbccf93ad66b&version=nightly
  - Catamorphism
  - Transducer: https://effectful.js.org/posts/simple-transducers-javascript.html

**Comment by mratsim** [comment]
7 years later. I have revisited this to build a high-performance AI inference library.

1. CuTe Layout Algebra from Nvidia is generic enough that you can reach with 2~3% of OpenBLAS performance, on CPU, with it: https://github.com/mratsim/tattletale/pull/35
2. You can build a full compiler as Nim macro with Cuda, OpenCL, Vulkan, WebGPU backends, run CuTe Layout Algebra or GEMM. I didn't have time to integrate Metal yet but it should be straightforward: https://github.com/mratsim/tattletale/pull/36

This means that Nim can be a DSL for GPU computing and kernels can be generated at compile-time and we avoid lowering issues like lowering ternaries on GPUs or handling builtins like min/max/abs.

Furthermore, if you look at exprgrad: https://github.com/can-lehmann/exprgrad/blob/905752c/benchmarks/matmul/matmul_gpu.nim (video: https://www.youtube.com/watch?v=YXD8ZODahts) it's a Halide-like DSL built on Nim: 

```nim

proc measure_cpu(gpu: GpuContext, a_tensor, b_tensor: Tensor[float32]) =
  let
    a = input("a", a_tensor.shape)
    b = input("b", b_tensor.shape)
  c*[y, x] ++= a[y, it] * b[it, x] | (x, y, it)
  let model = compile[float32](c.target("c"))
  
  measure "cpu":
    discard model.call("c", {"a": a_tensor, "b": b_tensor})

proc measure_naive(gpu: GpuContext, a_tensor, b_tensor: Tensor[float32]) =
  let
    a = input("a", a_tensor.shape)
    b = input("b", b_tensor.shape)
  c*[y, x] ++= a[y, it] * b[it, x] | (x, y, it)
  let model = compile[float32](c.target("c", CompileGpu), gpu=gpu)
  
  measure "naive":
    discard model.call("c", {"a": a_tensor, "b": b_tensor})

proc measure_tiled16(gpu: GpuContext, a_tensor, b_tensor: Tensor[float32]) =
  let
    a = input("a", a_tensor.shape)
    b = input("b", b_tensor.shape)
  c*[y, x] ++= a[y, it] * b[it, x] | (x, y, it) do:
    schedule:
      parallel(y)
      gpu:
        tile_size(x, 16)
        tile_size(y, 16)
        parallel(x)
        cache(a)
        cache(b)
        tile_size(it, 16)
        tile(it)
  let model = compile[float32](c.target("c", CompileGpu), gpu=gpu)
  
  measure "tiled16":
    discard model.call("c", {"a": a_tensor, "b": b_tensor})
```

For now I have deferred on the DSL approach because ML/DL/inference are moving targets and I'm afraid some layers appear (like RNN for Halide at first, possibly PagedAttention now?) that don't fit cleanly into a DSL or needs some annoying dance for max performance (Tensor cores and MMA, and Hopper TMAs, ...)

But it should be possible to translate exprgrad IR to my lower level GPU IR to have a DSL for fast iteration, possibly even exposing higher level construct CuteDSL or TileLang style, and the CuTe Layout Algebra library for the quest of highest performance.
